container/lxd/lxd.go

```diff
@@ -95,74 +135,278 @@ func (manager *containerManager) Namespace() instance.Namespace {
+	manager.server, err = ConnectLocal()
+	return err == nil
+}
 
-	// It is only possible to provision LXD containers
-	// of the same architecture as the host.
-	hostArch := arch.HostArch()
+func lxdSocketPath() string {
+	// LXD socket is different depending on installation method
+	// We prefer upstream's preference of snap installed LXD
+	debianSocket := filepath.FromSlash("/var/lib/lxd")
+	snapSocket := filepath.FromSlash("/var/snap/lxd/common/lxd")
+	path := os.Getenv("LXD_DIR")
+	if path != "" {
+		logger.Debugf("Using environment LXD_DIR as socket path: %q", path)
+	} else {
+		if _, err := os.Stat(snapSocket); err == nil {
+			logger.Debugf("Using LXD snap socket: %q", snapSocket)
+			path = snapSocket
+		} else {
+			logger.Debugf("LXD snap socket not found, falling back to debian socket: %q", debianSocket)
+			path = debianSocket
+		}
+	}
+	return filepath.Join(path, "unix.socket")
+}
 
-	hc = &instance.HardwareCharacteristics{AvailabilityZone: &manager.availabilityZone}
+// seriesLocalAlias returns the alias to assign to images for the
+// specified series. The alias is juju-specific, to support the
+// user supplying a customised image (e.g. CentOS with cloud-init).
+func seriesLocalAlias(series, arch string) string {
+	return fmt.Sprintf("juju/%s/%s", series, arch)
+}
 
-	imageName, err := manager.client.EnsureImageExists(
+// seriesRemoteAliases returns the aliases to look for in remotes.
+func seriesRemoteAliases(series, arch string) ([]string, error) {
+	seriesOS, err := jujuseries.GetOSFromSeries(series)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	switch seriesOS {
+	case jujuos.Ubuntu:
+		return []string{path.Join(series, arch)}, nil
+	case jujuos.CentOS:
+		if series == "centos7" && arch == jujuarch.AMD64 {
+			return []string{"centos/7/amd64"}, nil
+		}
+	case jujuos.OpenSUSE:
+		if series == "opensuseleap" && arch == jujuarch.AMD64 {
+			return []string{"opensuse/42.2/amd64"}, nil
+		}
+	}
+	return nil, errors.NotSupportedf("series %q", series)
+}
+
+// connectToSource connects to remote ImageServer using specified protocol.
+func (manager *containerManager) connectToSource(remote RemoteServer) (lxd.ImageServer, error) {
+	switch remote.Protocol {
+	case LXDProtocol:
+		return lxd.ConnectPublicLXD(remote.Host, &remote.ConnectionArgs)
+	case SimplestreamsProtocol:
+		return lxd.ConnectSimpleStreams(remote.Host, &remote.ConnectionArgs)
+	}
+	return nil, fmt.Errorf("Wrong protocol %s", remote.Protocol)
+}
+
+// startInstance starts previously created instance.
+func (manager *containerManager) startInstance(name string) error {
+	req := api.ContainerStatePut{
+		Action:  "start",
+		Timeout: -1,
+	}
+	op, err := manager.server.UpdateContainerState(name, req, "")
+	if err != nil {
+		return err
+	}
+	err = op.Wait()
+	return err
+}
+
+// stopInstance stops instance if it's not stopped.
+func (manager *containerManager) stopInstance(name string) error {
+	state, etag, err := manager.server.GetContainerState(name)
```

Comment:
I'm seeing etag used in multiple places.  What is it and why is it necessary?

Reply:
Quoting LXD docs:
To avoid race conditions, the Etag header should be read from the GET
response and sent as If-Match for the PUT request. This will cause LXD
to fail the request if the object was modified between GET and PUT.